src/time_utils.c

```diff
@@ -0,0 +1,127 @@
+/*
+ * This file and its contents are licensed under the Apache License 2.0.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-APACHE for a copy of the license.
+ */
+#include <postgres.h>
+#include <access/xact.h>
+#include <catalog/pg_type.h>
+#include <utils/timestamp.h>
+#include <utils/builtins.h>
+#include <parser/parse_coerce.h>
+#include <fmgr.h>
+
+#include "time_utils.h"
+#include "dimension.h"
+#include "utils.h"
+
+static Datum
+subtract_interval_from_now(Oid timetype, const Interval *interval)
+{
+	Datum res = DirectFunctionCall1(now, 0);
+
+	switch (timetype)
+	{
+		case TIMESTAMPOID:
+			res = DirectFunctionCall1(timestamptz_timestamp, res);
+			return DirectFunctionCall2(timestamp_mi_interval, res, IntervalPGetDatum(interval));
+		case TIMESTAMPTZOID:
+			return DirectFunctionCall2(timestamptz_mi_interval, res, IntervalPGetDatum(interval));
+		case DATEOID:
+			res = DirectFunctionCall1(timestamptz_timestamp, res);
+			res = DirectFunctionCall2(timestamp_mi_interval, res, IntervalPGetDatum(interval));
+			return DirectFunctionCall1(timestamp_date, res);
+		default:
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("unknown time type %s", format_type_be(timetype))));
```

Comment:
It is not possible with the current code since the only other supported (possible) time types are integer ones and this case is checked prior to calling this function. Note that this code was there before, I just moved it to a separate file so that it can be reused.